### PR TITLE
186 adding comments

### LINKED ISSUE #186: [186] Comment Section: Adding comments
**Current state**
User cannot add comments to a post

**Expected state**
User can add comments to a post

**To do**
Backend:  
- [x] Creating dao for comments with post functionality
- [x] Creating controller with routes
- [x] Creating commentservice
- [x] Testing route

Frontend:  
- [x] Creating comment component
- [x] Creating form for adding comment
- [x] Creating comment list
- [x] Creating comment item

**Branch**
This task will mostly be edited in the branch [186-adding-comments](/AjobK/Seaqull/tree/186-adding-comments)

### PR DESCRIPTION
**Pull request by MrGooss**
closes #186 

LEAVE AS IS, some necessary fixes have to be made before pulling

 **What is it supposed to do**
Add comments to posts

 **How can we test your branch**
Pull branch and add comment

 **What browser have been checked**
- [x] Chrome
- [ ] Firefox
- [ ] Opera
- [ ] Internet Explorer
- [x] Edge
- [ ] Safari

### CODE REVIEW

File: backend/app/service/commentService.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        if (req.params.path && foundComments)
+            return res.status(200).json(foundComments);
+        else
+            return res.status(404).json({ 'message': 'No comments found on that path' });
+    }
+
+    public createComment = async (req: Request | any, res: Response): Promise<Response> => {
+        const newComment = new comment()
+        const { path, content } = req.body
+        console.log('made it here')
```
Comment (AjobK) [inline backend/app/service/commentService.ts]: Please remove console logs

File: frontend/src/components/commentEditor/index.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        const totalLength = this.state.editorState.getCurrentContent().getPlainText().length + chars.length
+
+        return totalLength > this.maxLength
+    }
+
+    
+    onChange = (editorState) => {
+        this.setState({editorState});
+
+        console.log(convertToRaw(editorState.getCurrentContent()))
```
Comment (AjobK) [inline frontend/src/components/commentEditor/index.js]: Please remove console logs

File: frontend/src/components/commentForm/index.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            content: null
+        }
+
+        this.state = {
+            comment: this.comment
+        }
+    }
+
+    onCommentSubmit = () => {
+        console.log('saving')
```
Comment (AjobK) [inline frontend/src/components/commentForm/index.js]: Please remove console logs

File: frontend/src/components/commentForm/index.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        const url = `http://localhost:8000/api/comment/`
+
+        Axios.post(url, this.state.comment, {withCredentials: true}).then(response => {
+            console.log(response.data)
```
Comment (AjobK) [inline frontend/src/components/commentForm/index.js]: Please remove console logs

File: frontend/src/components/commentForm/index.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            this.resetComment()
+        }).catch(err => {
+            console.log(err)
```
Comment (AjobK) [inline frontend/src/components/commentForm/index.js]: Please remove console logs

File: frontend/src/components/commentSection/index.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        this.data = []
+    }
+
+    loadComments = () => {
+        let path = window.location.pathname.split('/').filter(i => i != '').pop();
+        const url = `http://localhost:8000/api/comment/${path}`
+
+        Axios.get(url)
+        .then(response => {
+            console.log(response.data)
```
Comment (AjobK) [inline frontend/src/components/commentSection/index.js]: Please remove console logs

File: frontend/src/components/comment/index.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
+import React, { Component } from 'react'
+import styles from './comment.scss'
+
+class Comment extends Component {
+
+    //TODO: move to util
```
Comment (AjobK) [inline frontend/src/components/comment/index.js]: TODO?

File: frontend/src/components/commentEditor/index.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
+import React, { Component, createRef } from 'react'
+import { EditorState, Editor, RichUtils, convertFromRaw, convertToRaw, ContentState } from 'draft-js'
+import "draft-js/dist/Draft.css";
```
Comment (AjobK) [inline frontend/src/components/commentEditor/index.js]: Please use single quotes

File: frontend/src/components/comment/index.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+                    return Math.round(diff / day) + ' month ago'
+
+                return Math.round(diff / month) + ' months ago'
+            case diff > year:
+                if (diff < 2 * year)
+                    return Math.round(diff / day) + ' year ago'
+
+                return Math.round(diff / year) + ' years ago'
+            default:
+                return ""
```
Comment (AjobK) [inline frontend/src/components/comment/index.js]: Please use single quotes

File: frontend/src/components/commentEditor/index.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        if (!this.maxLength) return false
+
+        const totalLength = this.state.editorState.getCurrentContent().getPlainText().length + chars.length
+
+        return totalLength > this.maxLength
+    }
+
+    
+    onChange = (editorState) => {
+        this.setState({editorState});
```
Comment (AjobK) [inline frontend/src/components/commentEditor/index.js]: Please remove semi-colons

File: frontend/src/components/commentSection/index.js [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+import { Section } from '../../layouts'
+
+class CommentSection extends Component {
+    constructor(props) {
+        super(props)
+        this.data = []
+    }
+
+    loadComments = () => {
+        let path = window.location.pathname.split('/').filter(i => i != '').pop();
```
Comment (AjobK) [inline frontend/src/components/commentSection/index.js]: Please remove semi-colons

File: backend/app/controllers/commentController.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
+import ControllerBase from '../interfaces/ControllerBase'
+import * as express from 'express'
+import CommentService from '../service/commentService'
+const isAuth = require('../middleware/isAuth');
```
Comment (AjobK) [inline backend/app/controllers/commentController.ts]: Semi-colon

File: backend/app/service/commentService.ts [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+        this.profileDAO = new ProfileDAO()
+    }
+
+    public getComments = async (req: Request, res: Response): Promise<Response> => {
+        const foundComments = await this.dao.getComments(req.params.path)
+
+        if (req.params.path && foundComments)
+            return res.status(200).json(foundComments);
+        else
+            return res.status(404).json({ 'message': 'No comments found on that path' });
```
Comment (AjobK) [inline backend/app/service/commentService.ts]: Semi-colons

File: frontend/src/components/comment/index.js [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+    //TODO: move to util
+    timeAgo = (prevDate) => {
+        const diff = Number(new Date()) - prevDate;
+        const minute = 60 * 1000;
+        const hour = minute * 60;
+        const day = hour * 24;
+        const month = day * 30;
+        const year = day * 365;
+        switch (true) {
```
Comment (AjobK) [inline frontend/src/components/comment/index.js]: Switch true...?

File: frontend/src/components/comment/index.js [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+            case diff < year:
+                if (diff < 2 * month)
+                    return Math.round(diff / day) + ' month ago'
+
+                return Math.round(diff / month) + ' months ago'
+            case diff > year:
+                if (diff < 2 * year)
+                    return Math.round(diff / day) + ' year ago'
+
+                return Math.round(diff / year) + ' years ago'
```
Comment (AjobK) [inline frontend/src/components/comment/index.js]: If possible please make a time utility class or something similar. This is a bit extensive.

Multiple WET lines too

File: frontend/src/components/commentEditor/index.js (hunk 8 above)
Comment (AjobK) [reply frontend/src/components/commentEditor/index.js]: And semi-colon

File: backend/app/controllers/commentController.ts [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+        this.initRoutes()
+    }
+
+    public initRoutes(): void {
+        this.router.get(this.comment + '/:path', this.commentService.getComments)
+        this.router.post(this.comment, isAuth, this.commentService.createComment)
+    }
+}
+
+export default CommentController
```
Comment (AjobK) [inline backend/app/controllers/commentController.ts:22]: Newline

File: backend/app/dao/commentDao.ts [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+    }
+
+    public async createComment(newComment: Comment): Promise<any> {
+        const repository = await DatabaseConnector.getRepository(this.repo)
+
+        return repository.save(newComment)
+    }
+}
+
+export default CommentDAO
```
Comment (AjobK) [inline backend/app/dao/commentDao.ts]: Newline

File: backend/app/service/commentService.ts [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+        newComment.created_at = new Date()
+        newComment.updated_at = new Date()
+
+        await this.dao.createComment(newComment)
+
+        return res.status(200).json({ message: 'Comment added!' })
+    }
+}
+
+export default CommentService
```
Comment (AjobK) [inline backend/app/service/commentService.ts]: Newline

File: frontend/src/components/comment/index.js [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+                            {this.props.comment.content}
+                        </div>
+                    </div>
+                </section>
+            </article>
+        )
+    }
+}
+
+export default Comment
```
Comment (AjobK) [inline frontend/src/components/comment/index.js]: Newline

File: frontend/src/components/commentEditor/index.js [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+                        onChange={this.onChange}
+                        placeholder="Leave a comment..."
+                    />
+                </div>
+            </div>
+        )
+    }
+}
+
+export default CommentEditor
```
Comment (AjobK) [inline frontend/src/components/commentEditor/index.js]: Newline

File: frontend/src/components/commentForm/index.js [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+                    }}
+                    value={this.state.comment.content} 
+                />
+                <button type="button" onClick={this.onCommentSubmit}>Save</button>
+            </div>
+        )
+    }
+}
+
+export default CommentForm;
```
Comment (AjobK) [inline frontend/src/components/commentForm/index.js]: Newline

File: frontend/src/components/commentSection/index.js [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+                <Section noTitle>
+                    <CommentForm />
+                    { this.displayComments() }
+                </Section>
+            </div>
+        )
+    }
+}
+
+export default CommentSection
```
Comment (AjobK) [inline frontend/src/components/commentSection/index.js]: Newline

**Review by AjobK — CHANGES_REQUESTED** [review]
Mainly semi-colons, console.logs and newlines among other things

---

COMMENT: User can see comment field when loggedout. Please redirect to login page when user clicks on 'save' and he/she gets error

BUG: Also getting error when adding comment and going back to post.
image

MISSING FEATURE?: Comment is not directly added when pressing 'save' succesfully

File: frontend/src/components/commentEditor/index.js (hunk 2 above)
Comment (MrGooss) [reply frontend/src/components/commentEditor/index.js]: Fixed

**Comment by MrGooss** [comment]
Implemented changes, please review @AjobK

**Pull request closed without merging**